dlib/serialize.h

```diff
@@ -1592,6 +2168,93 @@ namespace dlib
         }
     }
 
+// ----------------------------------------------------------------------------------------
+
+    template <typename T, typename deleter>
+    void serialize (
+        const std::unique_ptr<T, deleter>& item,
+        std::ostream& out
+    )
+    {
+        try
+        {
+            bool is_non_empty = item.operator bool();
+            serialize(is_non_empty, out);
+            if (is_non_empty)
+                serialize(*item, out);
+        }
+        catch (serialization_error& e)
+        {
+            throw serialization_error(e.info + "\n   while serializing an object of type std::unique_ptr");
+        }
+    }
+
+    template <typename T, typename deleter>
+    void deserialize (
+        std::unique_ptr<T, deleter>& item,
+        std::istream& in
+    )
+    {
+        try
+        {
+            //when deserializing unique_ptr, this is fresh state, so reset the pointers, even if item is non-empty
+            
+            bool is_non_empty;
+            deserialize(is_non_empty, in);
+            item.reset(is_non_empty ? new T() : nullptr);
+            
+            if (is_non_empty)
+                deserialize(*item, in);
+        }
+        catch (serialization_error& e)
+        {
+            throw serialization_error(e.info + "\n   while deserializing an object of type std::unique_ptr");
+        }
+    }
+
+// ----------------------------------------------------------------------------------------
+
+    template <typename T>
+    void serialize (
+        const std::shared_ptr<T>& item,
+        std::ostream& out
+    )
+    {
+        try
+        {
+            bool is_non_empty = item.operator bool();
```

Comment:
It's really weird to call `.operator bool()`.  Compare against `nullptr` or just assign to `bool` or explicitly cast to `bool` if you want.

Reply:
Yep ok. I did this just so i was 100% certain it was calling the ```bool()``` operator overload. I didn't want any surprises.

Reply:
Also i found this in ```dlib/dlib/tuple/tuple.h```:

```
funct(a.template get<i>());
```

So i thought maybe Davis doesn't mind weird notation too much.

Reply:
Well, `template` has to be there to compile.